text/0152-telemetry-schemas.md

```diff
@@ -0,0 +1,1159 @@
+backend if the backend expects to find that attribute by its name.
+
+Semantic conventions are an important part of this problem. These conventions
+define what names and values to use for span attributes, metric names and other
+fields. If semantic conventions are changed the existing implementations
+(telemetry source or consumers) need to be also changed correspondingly.
+Furthermore, to make things worse, the implementations of telemetry sources and
+implementations of telemetry consumers that work together and that depend on the
+changed semantic convention need to be changed simultaneously, otherwise such
+implementations will no longer work correctly together.
+
+Essentially there is a coupling between 3 parties: 1) OpenTelemetry semantic
+conventions, 2) telemetry sources and 3) telemetry consumers. The coupling
+complicates the independent evolution of these 3 parties.
+
+We recognize the following needs:
+
+- OpenTelemetry semantic conventions need to evolve over time. When conventions
+  are first defined, mistakes are possible and we may want to fix the mistakes
+  over time. We may also want to change conventions to re-group the attributes
+  into different namespaces as our understanding of the attribute taxonomy
+  improves.
+
+- Telemetry sources over time may want to change the schema of the telemetry
+  they emit. This may be because for example the semantic conventions evolved
+  and we want to make our telemetry match the newly introduced conventions.
+
+- In an observability system there may simultaneously exist telemetry sources
+  that produce data that conforms to different telemetry schemas because
+  different sources evolve at a different pace and are implemented and
+  controlled by different entities.
+
+- Telemetry consumers have a need to understand what schema a particular piece
+  of received telemetry confirms to. The consumers also need a way to be able to
+  interpret the telemetry data that uses different telemetry schemas.
+
+This document proposes a solution to these needs.
+
+## Solution Summary
+
+We believe that the 3 parties described above should be able to evolve
+independently over time, while continuously retaining the ability to correctly
+work together.
+
+Telemetry Schemas are central to how we make this possible. Here is a summary of
+the proposal:
+
+- We introduce a file format for defining telemetry schemas.
+
+- Telemetry schemas are versioned. Over time the schema may evolve and telemetry
+  sources may emit data confirming to newer versions of the schema.
+
+- Telemetry schemas explicitly define transformations that are necessary to
+  convert telemetry data between different versions of the schema, provided that
+  such conversions are possible. When conversions are not possible it
+  constitutes a breaking change between versions.
+
+- Telemetry schemas are identified by Schema URLs.
+
+- Telemetry sources (e.g. instrumentation libraries) should include a schema URL
```

Comment:
Not explicitly stated (AFAICT);  should the telemetry schema include the version number?

Reply:
Yes, it is defined later in the doc.